gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java

```diff
@@ -0,0 +1,296 @@
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.util;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.avro.AvroRuntimeException;
+import org.apache.avro.Schema;
+import org.apache.commons.lang3.StringUtils;
+import org.codehaus.jackson.JsonNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+public class AvroFlattener {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AvroFlattener.class);
+
+  /***
+   * Flatten the Schema to un-nest recursive Records (to make it optimal for ORC)
+   * @param schema Avro Schema to flatten
+   * @return Flattened Avro Schema
+   */
+  public static Schema flatten(Schema schema) {
+    // Validate input
+    Preconditions.checkNotNull(schema);
+
+    // Flatten schema
+    Schema flattenedSchema = flatten(schema, 0);
+
+    LOG.info("Original Schema : " + schema);
+    LOG.info("Flattened Schema: " + flattenedSchema);
+
+    return flattenedSchema;
+  }
+
+  private static Schema flatten(Schema schema, int level) {
+    // New schema
+    Schema flattenedSchema;
+
+    // Process all Schema Types
+    // (Primitives are simply cloned)
+    switch (schema.getType()) {
+      case ARRAY:
+        // Array might be an array of recursive Records, flatten them
+        flattenedSchema = Schema.createArray(flatten(schema.getElementType(), level+1));
+        break;
+      case BOOLEAN:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case BYTES:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case DOUBLE:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case ENUM:
+        flattenedSchema =
+            Schema.createEnum(schema.getName(), schema.getDoc(), schema.getNamespace(), schema.getEnumSymbols());
+        break;
+      case FIXED:
+        flattenedSchema =
+            Schema.createFixed(schema.getName(), schema.getDoc(), schema.getNamespace(), schema.getFixedSize());
+        break;
+      case FLOAT:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case INT:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case LONG:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case MAP:
+        flattenedSchema = Schema.createMap(flatten(schema.getValueType(), level+1));
+        break;
+      case NULL:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case RECORD:
+        List<Schema.Field> flattenedFields = new ArrayList<>();
+        if (null != schema.getFields() && schema.getFields().size() > 0) {
+          // Recursively flatten fields
+          for (Schema.Field oldField : schema.getFields()) {
+            LinkedList<String> lineage = new LinkedList<>();
```

Comment:
Shouldn't the lineage be a list of Schemas rather than just names. That give more flexibility, e.g. if you want to concatenate doc strings.